### ISSUE TITLE
Support features of Duo

### ISSUE THREAD

**Issue by frankwallis** [systemjs/plugin-css#20]
I have found that Duo/Component actually have the best css bundling support and I thought it may be helpful to list a couple of their features which could be incorporated into plugin-css:

- auto-prefixing

Automatically applies vendor prefixes. Actually this is a duo plugin and could be a separate plugin in systemjs or included as part of plugin-css

- Symbolic linking of assets

When bundling, Duo identifies all assets it finds in the css files, creates symbolic links from the bundle output directory to the original asset file, and rewrites the urls to point there. This means in production you wouldn't have to deliver the jspm_packages folder you just deliver the build folder and it contains everything needed to run the app.

- following imports

When deciding what to include in a bundle, Duo follows imports in css in the same way as it follows requires in javascript. This means you can control what goes in the css bundle entirely in the css files and there is clean separation of css & javascript.

Thanks

**Comment by guybedford** [comment]
@frankwallis thanks for these suggestions, it is really great to get this feedback.

* auto-prefixing. I'm certainly open to this, and it can potentially be a feature in the CSS plugin. Ideally this would be supported through operating on the same parse tree as clean css uses. Something like a shared parse tree through a project like https://github.com/postcss/postcss would make a lot of sense here.
* Symbolic linking of assets. This sounds interesting, and would be good to discuss further. What is the real benefit? Are you saying that most web-servers support symbolic links in a public folder into a non-public folder, is that how this would work?
* Following imports. I strongly disagree with having CSS imports following the same resolution process as the Node resolution. The browser sets the rules, not the server. With jspm, it's worth remembering that our goal is eventually not to bundle - everything should still work then.

**Comment by frankwallis** [comment]
- yes postcss does look really good and would come with lots of prebuilt plugins which would be awesome. It also neatly avoids the issue of how to compile sass in the browser (which I just ran into...). I guess it would replace clean-css and the systemjs plugin would provide a way of loading the postcss plugins?

- the symbolic linking is for a workflow where you bundle your app, and then zip it up and publish the zip archive. Because all the assets are accessible from /build you don't need to put the jspm_packages folder in the zip file which makes it smaller and more secure. In fact duo lets you choose whether to copy the files or symbolically link them to /build.

- I take your point. It's just something I thought would work and was a bit surprised when it didn't, but I understand the reasoning. It still seems a little odd to me to have to require css files in my javascript. I noticed that the angular-material override in the registry shims its own css file which is neat. I know you made a change recently to allow overrides to go in package.json, but is it possible for packages to provide their *own* override configuration in package.json, which would let me shim my index.css files?

**Comment by guybedford** [comment]
* I've created https://github.com/systemjs/plugin-css/issues/22 to track the conversion to using postCSS.
* I'm really not sold on the symbolic linking workflow. If users really want this functionality, we could create a "copyAssets" option or something and have all assets copied and renormalized into a folder in the build directory. I've created https://github.com/systemjs/plugin-css/issues/23 to track support for this.

Thanks for the suggestions, I think that closes this issue for now.
